payjoin-cli/src/app/v2/mod.rs

```diff
@@ -248,64 +237,183 @@ impl App {
         &self,
-        session: &mut Receiver<WithContext>,
+        session: Receiver<WithContext>,
+        persister: &ReceiverPersister,
     ) -> Result<Receiver<UncheckedProposal>> {
         let ohttp_relay = self
             .unwrap_relay_or_else_fetch(Some(session.pj_uri().extras.endpoint().clone()))
             .await?;
 
+        let mut session = session;
         loop {
             let (req, context) = session.extract_req(&ohttp_relay)?;
             println!("Polling receive request...");
             let ohttp_response = post_request(req).await?;
-            let proposal = session
+            let state_transition = session
                 .process_res(ohttp_response.bytes().await?.to_vec().as_slice(), context)
-                .map_err(|_| anyhow!("GET fallback failed"))?;
-            log::debug!("got response");
-            if let Some(proposal) = proposal {
-                break Ok(proposal);
+                .save(persister);
+            match state_transition {
+                Ok(OptionalTransitionOutcome::Progress(next_state)) => {
+                    return Ok(next_state);
+                }
+                Ok(OptionalTransitionOutcome::Stasis(current_state)) => {
+                    session = current_state;
+                    continue;
+                }
+                Err(e) => return Err(e.into()),
+            }
+        }
+    }
+
+    async fn process_receiver_session(
+        &self,
+        session: ReceiverTypeState,
+        persister: &ReceiverPersister,
+    ) -> Result<()> {
+        match session {
+            ReceiverTypeState::WithContext(proposal) => {
+                let proposal = self.read_from_directory(proposal, persister).await?;
+                self.check_proposal(proposal, persister).await
             }
+            ReceiverTypeState::UncheckedProposal(proposal) =>
+                self.check_proposal(proposal, persister).await,
+            ReceiverTypeState::MaybeInputsOwned(proposal) =>
+                self.check_inputs_not_owned(proposal, persister).await,
+            ReceiverTypeState::MaybeInputsSeen(proposal) =>
+                self.check_no_inputs_seen_before(proposal, persister).await,
+            ReceiverTypeState::OutputsUnknown(proposal) =>
+                self.identify_receiver_outputs(proposal, persister).await,
+            ReceiverTypeState::WantsOutputs(proposal) =>
+                self.commit_outputs(proposal, persister).await,
+            ReceiverTypeState::WantsInputs(proposal) =>
+                self.contribute_inputs(proposal, persister).await,
+            ReceiverTypeState::ProvisionalProposal(proposal) =>
+                self.finalize_proposal(proposal, persister).await,
+            ReceiverTypeState::PayjoinProposal(proposal) =>
+                self.send_payjoin_proposal(proposal, persister).await,
+            _ => Ok(()),
         }
     }
 
-    fn process_v2_proposal(
+    #[allow(clippy::incompatible_msrv)]
```

Comment:
Gotta open an issue to fix these incompatible_msrv exceptions. They're not MSRV compliant!